Prepare for GHC 9.2.

erikd opened the pull request:
## Related Issue

ADP-2679

## Summary

This PR:
- updates dependencies of `cardano-wallet` to prepare for building with GHC 9.2.
- adds basic support for the Conway era, such that existing test suites pass.

This PR does **NOT**:

- add CI support for GHC 9.2.
- add extra test suite coverage that's specific to Conway.

We will add an appropriate CI support for GHC 9.2 in a **_subsequent_** PR.

Anviking committed:
Upgrade to open-api 3.1.0

Anviking committed:
Workaround for "maximum recursion depth exceeded"

openapi-spec-validator specifications/api/swagger.yaml

erikd committed:
Delete TypeErrorSpec tests

These tests would not compile with ghc-9.2. Discussed this extensively
with Jonathan Knowles <jonathan.knowles@iohk.io>.

erikd committed:
Make it build with ghc-9.2

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -707,9 +706,13 @@ prop_signTransaction_addsExtraKeyWitnesses
-    arbitrary = genericArbitrary
-    shrink = genericShrink
+  arbitrary = liftM2 (:|) arbitrary arbitrary
+  shrink (x :| xs) = mapMaybe NE.nonEmpty . shrinkList shrink $ x : xs
```
Suggestion: use `genNonEmpty` and `shrinkNonEmpty` from `Test.QuickCheck.Extra`

```suggestion
  arbitrary = genNonEmpty arbitrary
  shrink = shrinkNonEmpty shrink
```

jonathanknowles commented on lib/wallet/api/http/Cardano/Wallet/Launch/Cluster.hs:
```diff
@@ -577,16 +577,16 @@ configurePool tr baseDir era metadataServer recipe = do
 
             let updateStaking sgs = sgs
-                    { Ledger.sgsPools = (Map.singleton poolId params)
+                    { Ledger.sgsPools = (ListMap.fromList [(poolId, params)])
```
Just curious, is it possible to write:

```suggestion
                    { Ledger.sgsPools = (ListMap [(poolId, params)])
```

erikd replied:
I find the `fromList` version more obvious.

jonathanknowles commented on lib/wallet/cardano-wallet.cabal:
```diff
@@ -173,7 +174,7 @@ library
     , stm
     , streaming-commons
-    , strict-containers
+    , cardano-strict-containers
```
Finally! 💯 

This means we can now depend upon the hackage version of `strict-containers` (if we need to).

jonathanknowles commented on lib/wallet/src/Cardano/Api/Gen.hs:
```diff
@@ -1013,13 +1016,16 @@ genEpochNo :: Gen EpochNo
+                                   <*> mapM (const $ chooseInteger (0, 5_000)) costModelParams
+  case eCostModel of
+    Left err -> error $ "genCostModel: " ++ show err
+    Right cModel -> return . CostModel $ Alonzo.getCostModelParams cModel
```
Just some minor formatting fixes (our house style favours newline + 4-space-indent instead of vertical alignment).

```suggestion
    let costModelParams = Alonzo.getCostModelParams Plutus.testingCostModelV1
    eCostModel <- Alonzo.mkCostModel
        <$> genPlutusLanguage
        <*> mapM (const $ chooseInteger (0, 5_000)) costModelParams
    case eCostModel of
        Left err -> error $ "genCostModel: " ++ show err
        Right cModel -> return . CostModel $ Alonzo.getCostModelParams cModel
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Allegra.hs:
```diff
@@ -23,6 +23,19 @@ module Cardano.Wallet.Read.Primitive.Tx.Allegra
+import Control.Lens
+    ( (^.) )
+
+import Cardano.Ledger.Core
```
This will allow `stylish-haskell` to re-order the imports properly:

```suggestion
    ( (^.) )
import Cardano.Ledger.Core
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Mary.hs:
```diff
@@ -15,8 +15,21 @@ module Cardano.Wallet.Read.Primitive.Tx.Mary
+import Control.Lens
+    ( (^.) )
+
 import Cardano.Api
```
```suggestion
    ( (^.) )
import Cardano.Api
```

jonathanknowles reviewed (commented):
Hi @erikd 

Many thanks for making this PR!

I've had a brief look, and have some initial (small) suggestions.

Will take another look later on, when I get some time.

Thanks again!
Jonathan

Anviking commented on cabal.project:
```diff
@@ -91,76 +92,71 @@ source-repository-package
 
 -- It seems that if we remove this source-repository-package to rely on CHaP,
 -- for some reason, we're unable to retrieve the cardano-node executables in
 -- our nix/haskell.nix file.
```
It seems all of this should be removed then:
```haskell
-- -- Different version than from CHaP to include the ReaderT fix.
 -- -- TODO [ADP-2447] This source-repository-package should be removable with next node bump.
 -- source-repository-package
 --   type: git
 --   location: https://github.com/input-output-hk/io-sim
 --   tag: 1cdeed83b4cccf4ed1e656375200c889ecd879c7
 --   --sha256: 1wy33hs5rg2n0haf5wc9s9mrrkdzyp95njkkpgs8g5mnllapdka8
 --   subdir:
 --     io-classes
 --     io-sim
 --     strict-stm

 -- It seems that if we remove this source-repository-package to rely on CHaP,
 -- for some reason, we're unable to retrieve the cardano-node executables in
 -- our nix/haskell.nix file.
```

erikd replied:
I removed the `io-sim` part of that comment block.

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -707,9 +706,13 @@ prop_signTransaction_addsExtraKeyWitnesses
+-- instance Arbitrary a => Arbitrary (NonEmpty a) where
+--    arbitrary = genericArbitrary
+--    shrink = genericShrink
+
```
```suggestion
```

erikd committed:
Update dependencies

Most specifically update cardano-ledger which has changed significantly
and required a HUGE number of changes.

Co-Authored-By: Marcin Szamotulski <coot@coot.me>
Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>
Co-authored-by: Johannes Lund <johannes.lund@iohk.io>
Co-authored-by: Paolo Veronelli <paolo.veronelli@gmail.com>

erikd committed:
Update dependencies

Most specifically update cardano-ledger which has changed significantly
and required a HUGE number of changes.

Co-Authored-By: Marcin Szamotulski <coot@coot.me>
Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>
Co-authored-by: Johannes Lund <johannes.lund@iohk.io>
Co-authored-by: Paolo Veronelli <paolo.veronelli@gmail.com>

jonathanknowles committed:
Minor formatting fixes.

- Use 4-space indents.
- 80-column line limit.

(See https://input-output-hk.github.io/adrestia/code/Coding-Standards)

Anviking committed:
Add extra 'ErrMkTransactionTokenQuantityExceedsLimit'

duplicating coin-selection validation. C.f. code comment.

erikd committed:
Add basic support for the Conway era.

andreabedini committed:
Bump hackage.nix

andreabedini committed:
Bump CHaP

jonathanknowles committed:
Fix functions that cast from `RecentEraTxOut` to specific eras.

andreabedini committed:
Bump haskell.nix to support revisions

andreabedini committed:
Workaround for broken package

jonathanknowles committed:
Fix `shrinkBinaryData` shrinker.

jonathanknowles committed:
Remove rebase conflicts. (Note, this doesn't make the code compile.)

jonathanknowles committed:
Fix compile errors in `Shelley.Compatibility`.

jonathanknowles committed:
Remove `V1` versus `V2` distinction for simple scripts.

jonathanknowles committed:
Fix compilation errors in `toLedgerScript`.

jonathanknowles committed:
Fix compilation errors in `Shelley.Transaction`.

jonathanknowles committed:
Add dependency on `ouroboros-network-protocols`.

jonathanknowles committed:
Add missing constraints within `Write.Tx`.

jonathanknowles committed:
Specify `IsRecentEra` constraint in `buildCoinSelectionForTransaction`.

We only support building transactions in "recent eras", and
"IsRecentEra" already provides us with the superclass constraint that
was missing, so using it kills two birds with one stone.

jonathanknowles committed:
Add dependency on `ouroboros-consensus-diffusion`.

jonathanknowles committed:
Add missing argument to `codecConfig`.

jonathanknowles committed:
Fix compilation errors in `Launch.Cluster`.

jonathanknowles committed:
Revert to using `createPool` in `DB.Sqlite`.

This function is deprecated, and we should fix it, but that is outside
the scope of this PR.

jonathanknowles committed:
Add dependency on `ouroboros-network-api`.

jonathanknowles committed:
Fix compilation errors in `Api.GenSpec`.

jonathanknowles committed:
Fix errors in `Shelley.CompatibilitySpec`.

Anviking committed:
Drop now unneeded normalizeCardanoUTxO

Fixes one compilation error.

Anviking committed:
Fixes to Write.TxSpec

Anviking committed:
Fixes to Write.TxSpec & Shelley.TransactionSpec

Anviking committed:
Complete removal of SimpleScript version distinction in CompatibilitySpec

jonathanknowles committed:
Disable deprecation warnings for `Cardano.DB.Sqlite`.

Tracked in the following ticket:
https://input-output.atlassian.net/browse/ADP-2841

jonathanknowles committed:
Disable deprecation warnings in specific modules.

Tracking ticket: https://input-output.atlassian.net/browse/ADP-2841

jonathanknowles committed:
Disable ambiguous field warnings in specific modules.

Tracking ticket: https://input-output.atlassian.net/browse/ADP-2841

jonathanknowles committed:
Use `AlonzoScript` instead of `Alonzo.Script`.

(The latter is deprecated.)

jonathanknowles committed:
Disable redundant constraint warnings in specific modules.

Tracking ticket: https://input-output.atlassian.net/browse/ADP-2841

jonathanknowles committed:
Remove a whole set of redundant constraints.

jonathanknowles committed:
Remove all remaining redundant imports.

jonathanknowles committed:
Remove a completely redundant pattern match.

No idea how this was added. But it'll never match, so we can remove it.

jonathanknowles committed:
Disable "star-is-type" warnings in specific modules.

Tracking ticket: https://input-output.atlassian.net/browse/ADP-2841

jonathanknowles committed:
Fix incomplete pattern match warning in `UTxOIndex.checkInvariant`.

Anviking committed:
Fix scriptToCardanoEnvelopeJSON roundtrip

Anviking committed:
Fix some warnings mostly in Shelley.TransactionSpec

Unisay committed:
Fix partial pattern matches by adding Conway case

erikd committed:
Add further support for the Conway era.

erikd committed:
Remove some deprecation warnings

jonathanknowles committed:
Replace line continuation characters with `mconcat` and `unwords`.

Line continuation characters don't work in the presence of CPP.

But we need to use CPP to mark certain sections of the code as
conditional on certain GHC versions.

So we replace these line continuations with `mconcat` and `unwords`.

jonathanknowles committed:
Make `-fno-warn-ambiguous-fields` conditional on the GHC version.

It's not possible to disable this type of warning in GHC 8.10.7, because
this type of warning was not yet invented at the time GHC 8.10.7 was
released.

However, we do need to disable these warnings in GHC 9.2.1
(temporarily), so we must make the "disabling of this warning type"
conditional on the compiler version.

We will eventually fix these warnings. See the following tracking
ticket:

https://input-output.atlassian.net/browse/ADP-2841

jonathanknowles committed:
Appease `stylish-haskell`.

jonathanknowles committed:
Remove use of unused `TypeApplications` extension in `ReferenceInputs`.

Unisay committed:
update: github action to publish docs uses newer versions

Anviking committed:
Add ConwayGenesis to integration test cluster

Unisay committed:
openapi-spec-validator uses schema 3.1.0

Anviking committed:
Workaround pingPong_2 test data broken by CPP

Anviking committed:
Fix pingPong_1 in TransactionSpec

Anviking committed:
Fixup: hlint

Anviking committed:
Restrict signing tests to recent eras

Quick workaround for failure in the Shelley era. The proper solution is
to also restrict the implementation to only work in recent eras, for
which a JIRA task was created.

Techincally, hiding this failure is a API breaking change, but it would
be /highly/, highly unlikely for any user to care about signing txs in
the Shelley era... with non-compatibile timelock scripts. Besides, the
preceding constructTx endpoint only works in recent eras.

Anviking committed:
Add pingPong_2 and testPParam binary goldens

Generated on master. Also, fix the CostModels.

jonathanknowles committed:
Bump version of `iohk-nix`.

Steps I followed:

Find the latest revision of `iohk-nix` at:

https://github.com/input-output-hk/iohk-nix/commits/

Use that to update the revision of iohkNix in flake.lock:

```diff
diff --git a/flake.lock b/flake.lock
index 8062f9357..7d0279b23 100644
--- a/flake.lock
+++ b/flake.lock
@@ -635,7 +635,7 @@
         "narHash": "sha256-YhS7zGd6jK/QM/+wWyj0zUBZmE3HOXAL/kpJptGYIWg=",
         "owner": "input-output-hk",
         "repo": "iohk-nix",
-        "rev": "7fc7625a9ab2ba137bc70ddbc89a13d3fdb78c8b",
+        "rev": "718a10193b92f5bb2c0d2f068273a6d6d8c46034",
         "type": "github"
       },
       "original": {
```

Run:

```
nix flake lock --update-input iohkNix
```

Then:

```
nix/regenerate.sh
```

paolino committed:
fix parseEraValue to cover Conway

Unisay committed:
Revert "Upgrade to open-api 3.1.0"

This reverts commit 4cf7c0cd707d117d5502d9069636ebdbacd3d0c3.

Unisay committed:
Revert "openapi-spec-validator uses schema 3.1.0"

This reverts commit 8c6b1600995ed20853aa4a62e9eefcb574a0003a.

Unisay committed:
swagger.yaml includes conway era

jonathanknowles committed:
Remove `cabalWrapped`.

On the advice of:

- Andrea Bedini
- Moritz Angermann

jonathanknowles committed:
Temporarily disable `systemd` flags for `cardano-{config,node}`.

Otherwise the build fails with:

```
$ cabal build all
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: cardano-node-1.36.0 (user goal)
[__1] trying: cardano-node:+systemd
[__2] trying: lobemo-scribe-systemd-0.1.0.0 (dependency of cardano-node
+systemd)
[__3] next goal: libsystemd-journal (dependency of lobemo-scribe-systemd)
[__3] rejecting: libsystemd-journal-1.4.5, libsystemd-journal-1.4.4 (conflict:
pkg-config package libsystemd>=209, not found in the pkg-config database)
[__3] rejecting: libsystemd-journal-1.4.3, libsystemd-journal-1.4.2,
libsystemd-journal-1.4.1, libsystemd-journal-1.4.0, libsystemd-journal-1.3.4,
libsystemd-journal-1.3.3, libsystemd-journal-1.3.1, libsystemd-journal-1.3.0,
libsystemd-journal-1.2.0, libsystemd-journal-1.1.0, libsystemd-journal-1.0.0
(constraint from project config
/home/jsk/projects/input-output-hk/cardano-wallet/cabal.project requires
>=1.4.4)
[__3] fail (backjumping, conflict set: libsystemd-journal,
lobemo-scribe-systemd)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: libsystemd-journal, cardano-node,
cardano-node:systemd, lobemo-scribe-systemd
```

Unisay committed:
add `cabal` tool into the nix shell

Unisay committed:
Use nixpkgs-unstable for the openapi-spec-validator dependency

Unisay committed:
Merge pull request #3798 from input-output-hk/yura/ghc-9.2

Use nixpkgs-unstable for the openapi-spec-validator dependency

Anviking commented on cabal.project:
```diff
@@ -202,54 +161,55 @@ constraints:
+  flags: -systemd
 package cardano-node
-  flags: +systemd
+  flags: -systemd
```
Is this ok to change?

jonathanknowles replied:
Without this change, I could not get this branch to build (as cabal was not able to satisfy the dependency). I honestly don't know what effect it will have, if any.

angerman replied:
It just makes cardano-node not depend on systemd for logging. I'd argue depending on systemd is wrong to begin with though. Using it is a different topic but having a hard dependency on it?

All static node builds have this disabled as well. It has no functional impact.
